### PR TITLE
libtest: include test output in junit xml reports

### LINKED ISSUE #110336: `--format=junit` doesn't include stdout / test failure reason
I tried this code:

```rust
#[test]
fn it_breaks() { panic!("PANICCC"); }
```

I expected to see this happen: the string `PANICCC` in the test output, when running `cargo test -- --format=junit -Zunstable-options`.

Instead, this happened:

```xml
<?xml version="1.0" encoding="UTF-8"?><testsuites><testsuite name="test" package="test" id="0" errors="0" failures="1" tests="1" skipped="0" ><testcase classname="crate" name="it_breaks" time="0"><failure type="assert"/></testcase><system-out/><system-err/></testsuite></testsuites>
```

### Meta

`rustc --version --verbose`:
```
rustc 1.68.2 (9eb3afe9e 2023-03-27)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: x86_64-unknown-linux-gnu
release: 1.68.2
LLVM version: 15.0.6
```

</p>
</details>

### PR DESCRIPTION
**Pull request by durin42**
Fixes #110336.

### CODE REVIEW

File: tests/run-make/libtest-junit/output-default.xml [hunk 1]
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="UTF-8"?><testsuites><testsuite name="test" package="test" id="0" errors="0" failures="1" tests="4" skipped="1" ><testcase classname="unknown" name="a" time="$TIME"><system-out><![CDATA[print from successful test]]>&#xA;<![CDATA[]]></system-out></testcase><testcase classname="unknown" name="b" time="$TIME"><failure type="assert"/><system-out><![CDATA[print from failing test]]>&#xA;<![CDATA[thread 'b' panicked at 'assertion failed: false', f.rs:10:5]]>&#xA;<![CDATA[note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace]]>&#xA;<![CDATA[]]></system-out></testcase><testcase classname="unknown" name="c" time="$TIME"><system-out><![CDATA[thread 'c' panicked at 'assertion failed: false', f.rs:16:5]]>&#xA;<![CDATA[]]></system-out></testcase><system-out/><system-err/></testsuite></testsuites>
```
Comment (cuviper) [inline tests/run-make/libtest-junit/output-default.xml:1]: Why do "a" and "c" still have their output here? (without `--show-output`)

File: library/test/src/formatters/junit.rs [hunk 2]
```diff
@@ -26,6 +26,18 @@ impl<T: Write> JunitFormatter<T> {
     }
 }
 
+fn str_to_cdata(s: &str) -> String {
+    // Drop the stdout in a cdata. Unfortunately, you can't put either of `]]>` or
+    // `<?'` in a CDATA block, so the escaping gets a little weird.
+    let escaped_output = s.replace("]]>", "]]]]><![CDATA[>");
+    let escaped_output = escaped_output.replace("<?", "<]]><![CDATA[?");
+    // We also smuggle newlines as &#xa so as to keep all the output on line line
```
[the remark this answers is not in the document]
Comment (durin42) [reply library/test/src/formatters/junit.rs]: Whoops fixed

File: tests/run-make/libtest-junit/output-default.xml (hunk 1 above)
Comment (durin42) [reply tests/run-make/libtest-junit/output-default.xml]: Ah, nuts. Good catch, fixed.

File: library/test/src/formatters/junit.rs [hunk 3]
```diff
@@ -26,6 +26,18 @@ impl<T: Write> JunitFormatter<T> {
     }
 }
 
+fn str_to_cdata(s: &str) -> String {
+    // Drop the stdout in a cdata. Unfortunately, you can't put either of `]]>` or
+    // `<?'` in a CDATA block, so the escaping gets a little weird.
+    let escaped_output = s.replace("]]>", "]]]]><![CDATA[>");
+    let escaped_output = escaped_output.replace("<?", "<]]><![CDATA[?");
+    // We also smuggle newlines as &#xa so as to keep all the output on one line
+    let escaped_output = escaped_output.replace("\n", "]]>&#xA;<![CDATA[");
```
Comment (riking) [inline library/test/src/formatters/junit.rs:35]: The compiler already pulls in aho-corasick, but the test library does not. Failing test stdout could easily balloon to gigantic sizes -- should I write a PR to swap these replaces to use aho-corasick?

Comment (cuviper) [reply library/test/src/formatters/junit.rs:35]: I don't think so -- it's a higher bar to add dependencies to the standard library, including `test`.

Comment (riking) [reply library/test/src/formatters/junit.rs:35]: discussion on Zulip: Let's wait until someone does a perf measurement and finds this to blame.

**Pull request merged**